### DISCUSSION TITLE
Linux vscode / cmake integration

### DISCUSSION THREAD

**Discussion by hros** [microsoft/vcpkg#24491] [Q&A]
I followed the instructions in this [blog](https://www.40tude.fr/how-to-use-vcpkg-with-vscode-and-cmake/), and it seems that I have to manually add each time I create a project the lines:
```
include_directories(
    /abs_path_to_vcpkg/installed/x64-linux/include
)
link_directories(
    /abs_path_to_vcpkg/installed/x64-linux/lib
)
```
is there a way to have these settings added automatically whenever a new cmake project is created?

**Comment by Osyotr** [comment]
Use `find_package` instead.

**Reply by hros** [reply]
Thanks for your reply. 
Please provide an example.
does `find package` add include and library directories to the project?
My project uses `sqlite3` which is installed by `vcpkg`, but just adding `find_package(sqlite3)` does not work. 
I get the following output from cmake:
```
[cmake]   By not providing "Findsqlite3.cmake" in CMAKE_MODULE_PATH this project has
[cmake]   asked CMake to find a package configuration file provided by "sqlite3", but
[cmake]   CMake did not find one.
[cmake] 
[cmake]   Could not find a package configuration file provided by "sqlite3" with any
[cmake]   of the following names:
[cmake] 
[cmake]     sqlite3Config.cmake
[cmake]     sqlite3-config.cmake
[cmake] 
[cmake]   Add the installation prefix of "sqlite3" to CMAKE_PREFIX_PATH or set
[cmake]   "sqlite3_DIR" to a directory containing one of the above files.  If
[cmake]   "sqlite3" provides a separate development package or SDK, be sure it has
[cmake]   been installed.
```

**Reply by Osyotr** [reply]
If the package exports its configuration or CMake has built-in find-module for this package or it's a header-only library, you may notice something like this in the output log after running `vcpkg install`:
```
The package sqlite3:x86-windows provides CMake targets:

    find_package(unofficial-sqlite3 CONFIG REQUIRED)
    target_link_libraries(main PRIVATE unofficial::sqlite3::sqlite3)
```
Otherwise, you should find the library yourself: https://github.com/microsoft/vcpkg/blob/master/docs/examples/installing-and-using-packages.md#handling-libraries-without-native-cmake-support

**Reply by dg0yt** [reply]
In particular for sqlite3, you would normally use CMake's provided Find module [`FindSQLite3.cmake`](https://cmake.org/cmake/help/latest/module/FindSQLite3.html) - and you must match the case for Find modules:
~~~cmake
find_package(SQLite3)
~~~

Neither config or nor Find module automatically adds include dirs or link libraries. They will provides result variables and/or targets which you can refer to in CMake code. If available, targets are the easiest way to pass all usage requirements to your CMake targets, via `target_link_libraries`.

**Reply by hros** [reply]
> If the package exports its configuration or CMake has built-in find-module for this package or it's a header-only library, you may notice something like this in the output log after running `vcpkg install`:
> 
> ```
> The package sqlite3:x86-windows provides CMake targets:
> 
>     find_package(unofficial-sqlite3 CONFIG REQUIRED)
>     target_link_libraries(main PRIVATE unofficial::sqlite3::sqlite3)
> ```
> 
> Otherwise, you should find the library yourself: https://github.com/microsoft/vcpkg/blob/master/docs/examples/installing-and-using-packages.md#handling-libraries-without-native-cmake-support

using the output of `vcpkg install pkg` and (almost blindly) copying it to the `CMakeLists.txt` file is the no-nonsense way I was looking for, simply works.
could be further automated by the vscode plugin, by getting the list of installed packages from the output of `vcpkg list`, and when the user selects a package, the extension could automatically add the output of running `vcpkg install pkg`

While I imagine further improvements, wouldn't it be terrific if the vscide extensions could suggest as a fix to an unknown #include (import) of a header (module) to automatically add the appropriate vcpkg from the installed list.....

Perhaps even add a cleanup action removing from the cmake configuration unused packages

**Reply by dg0yt** [reply]
> using the output of `vcpkg install pkg` and (almost blindly) copying it to the `CMakeLists.txt` file is the no-nonsense way I was looking for, simply works.

It works best when the maintainer read the upstream documentation (or CMake documentation) and manually wrote the small usage snippet.
It doesn't always work for heuristical output: https://github.com/microsoft/vcpkg/issues/20190
